Add prepare-disk command.

kk-src opened the pull request:
Prepare-disk command formats a given disk with ext4. This is required for a disk(as passthrough) to be made available within a container. Here is an excerpt from a configuration for such a scenario.

"mounts": [
     {
       "host_path": "\\\\.\\PHYSICALDRIVE2",
       "container_path": "/disk"
     }

The gcs/runc combo expects the pass-through disk to be formatted and tries to mount it on a local path within
the uvm. Prepare-disk ensures that this restriction is met.

Note: Full disk is formatted with ext4 without any partitioning. TODO: add options to allow partitioning and then formatting.

kk-src commented:
/cc @kevpar

dcantah commented on internal/lcow/common.go:
```diff
@@ -0,0 +1,64 @@
+
+// FormatDiskUvm creates a utility vm, mounts the disk as a scsi disk onto to the VM
+// and then formats it with ext4.
+func FormatDiskUvm(ctx context.Context, lcowUVM *uvm.UtilityVM, controller int, lun int32, destPath string) error {
```
Does this need to be exported? Seems its only used in FormatDisk which is whats actually called in runhcs

kk-src replied:
Thanks @dcantah. Yep, not required outside the package. We can use the same for scratch space as well, but that is also in lcow package. Updated in the latest iteration.

dcantah commented:
And the CI is failing because we just recently moved hcsoci.CommandContext to a new /internal/cmd package.

kk-src commented:
>The CI is failing because we just recently moved hcsoci.CommandContext to a new /internal/cmd package.

Thank you @dcantah . Updated the PR with the latest package used.

dcantah commented on internal/lcow/disk.go:
```diff
@@ -0,0 +1,55 @@
+	}
+	removeSCSI := true
+	defer func() {
+		if removeSCSI {
```
Was this supposed to be set to false anywhere for some codepath to avoid cleanup? Right now it gets set to true above and then not changed anywhere else (unless I'm missing it 😄)

dcantah commented on internal/lcow/disk.go:
```diff
@@ -0,0 +1,55 @@
+	removeSCSI := true
+	defer func() {
+		if removeSCSI {
+			lcowUVM.RemoveSCSI(ctx, destPath)
```
Nit: Prefer `scsi.Release(ctx)` instead.

dcantah commented on internal/lcow/disk.go:
```diff
@@ -0,0 +1,55 @@
+// can be used to offline the disk.
+func FormatDisk(ctx context.Context, lcowUVM *uvm.UtilityVM, destPath string) error {
+	if lcowUVM == nil {
+		return fmt.Errorf("no uvm")
```
Nit: Prefer `errors.New()` as there isn't any formatting used here.

kk-src replied:
Keeping it as to keep it similar to the rest of the code in the repo - e.g. disk.go and others.

dcantah commented on internal/lcow/disk.go:
```diff
@@ -0,0 +1,55 @@
+		"lun":        scsi.LUN,
+	}).Debug("lcow::FormatDisk device attached")
+
+	err = formatDiskUvm(ctx, lcowUVM, scsi.Controller, scsi.LUN, destPath)
```
Can be 
```go
if err := formatDiskUvm(ctx, lcowUVM, scsi.Controller, scsi.LUN, destPath); err != nil {
          return err
}
```

dcantah commented on internal/lcow/common.go:
```diff
@@ -0,0 +1,64 @@
+	mkfsCtx, cancel := context.WithTimeout(ctx, timeout.ExternalCommandToStart)
+	cmd = cmdpkg.CommandContext(mkfsCtx, lcowUVM, "mkfs.ext4", "-q", "-E", "lazy_itable_init=0,nodiscard", "-O", `^has_journal,sparse_super2,^resize_inode`, device)
+	var mkfsStderr bytes.Buffer
+	cmd.Stderr = &mkfsStderr
```
It looks like we purposely set stderr to be output to a buffer but then we don't use this anywhere afterwards

dcantah commented:
@kk-src Thanks for the PR :) Left some comments. Also could you sign-off the commit with `git commit --amend -s`

kk-src commented:
Thank you so much @dcantah. Some of this is existing code, just tried to move to a common code block :-) Neverthless, will get back after addressing the comments and signing off.

dcantah commented:
@kk-src Awesome sounds good

kk-src commented:
@dcantah - address comments, signed the commit, PTAL.

dcantah commented on internal/lcow/disk.go:
```diff
@@ -0,0 +1,55 @@
+	if err != nil {
+		return err
+	}
+	removeSCSI = true
+	defer func() {
+		if removeSCSI {
+			scsi.Release(ctx)
+		}
+	}()
```
Still not understanding this as if we reached this point then removeSCSI is always true so we'll always run the condition in the defer. What's the reason for the variable?

kk-src replied:
Apologies! Removed the unnecessary check.

dcantah reviewed (approved):
LGTM!

kevpar reviewed (approved):
LGTM
